language/move-vm/runtime/src/interpreter.rs

```diff
@@ -1000,21 +1025,44 @@ impl Frame {
                             interpreter.operand_stack.last_n(field_count as usize)?,
                         )?;
                         let args = interpreter.operand_stack.popn(field_count)?;
-                        interpreter
-                            .operand_stack
-                            .push(Value::struct_(Struct::pack(args)))?;
+                        let value = Value::struct_(Struct::pack_with_tag(
+                            args,
+                            resolver
+                                .instantiate_generic_type(*si_idx, self.ty_args())?
+                                .get_hash(),
+                        ));
+
+                        interpreter.operand_stack.push(value)?;
```

Comment:
We need to evaluate the perf implications this might have very carefully. `instantiate_generic_type` is a very expensive operation and by calling it we're also making pack/unpack no longer O(1)

Reply:
We may be able to cache this computation because the tag should remain unchanged for this instruction

Reply:
This is actually the thing might be concerning, cuz we won't be able to avoid this check.